model_discretization_api.py

```diff
@@ -0,0 +1,209 @@
+# Delta2D, Box2D, ... should have the possibility to apply a different normalization,
+# where the amplitude parameter corresponds to the integral of the model and not the
+# peak value at maximum (see e.g.: https://gammapy.readthedocs.org/en/latest/api/gammapy.morphology.Shell2D.html#gammapy.morphology.Shell2D)
+# This is very useful for modeling sources, because the amplitude parameter then
+# corresponds to the total flux of the source...
+model = Gaussian2D(1, 0, 0, 1, 1, normalization='peak')
+model = Gaussian2D(1, 0, 0, 1, 1, normalization='integral')
+
+# alternatively one could introduce new NormGaussian2D, NormBox2D, ... models, but that
+# would probably be a lot of code duplication?
+
+# all models that are 'normalisable' should have an `extent` attribute that specifies
+# a rectangular region where the model is nonzero or has 99.99% containment (or a similar
+# criterion). A reasonable value would be chosen as default and would be given as multiples 
+# model's width parameters. Model evaluation can later be limited to the region defined
+# by `extent` for better performance 
+Gaussian2D.extent= dict(x=5 * x_stddev, y=5 * y_stddev)
+Disk2D.extent = dict(x=R_0, y=R_0) 
+
+# models should have an attribute that defines a reasonable sample size for the model,
+# warnings could be raised, when models are undersampled.
+Gaussian2D.sample_size = width / 5.
+
+
+###############################
+# Model Image Computation API #
+###############################
+
+# functionality is added to NDData via NDDataMixin classes
+class SkyImage(NDData, NDSlicingMixin, NDWCSCutoutMixin) : pass
+
+# a sky image object is created as following
+sky_image = SkyImage(np.zeros(501, 10001), wcs=WCS())
+
+# it has a coordinates attribute, that returns a SkyCoord object
+# with every pixel containing the corresponding position
+sky_image.coordinates
+
+# one can make rectangular cutouts of the sky image using
+cutout = sky_image.extract(position, extent)
+cutout = sky_image.extract(postion, extent, copy=True)
+
+# which returns a SkyImage object, but with modified WCS transform 
+# data can be copied optionally
+
+# to render the model on a sky image a new function should be defined
+source_image = render_model_to_sky(model, wcs=WCS())
+
+# which would return again a SkyImage object centered on the nearest pixel of the 
+# source position of size defined by extent. The pixel values are computed according to
+# the model's mode. 
+
+# The source image can then be added to sky_image 
+sky_image.add(source_image)
+
+# later `.add` could use any kind of resampling/reprojection
+# for now we just assume that `sky_image` and `source_image`
+# pixels are aligned (which should be checked...)
+
+# Combining it all, 
+sky_image.add_sources(models)
+
+# models should be an iterable of model instances for each source.
+# then paralleization can speed things up.
+
+sky_image.add_sources(models, distribute=True, N_core='none')
```

Comment:
@embray – There is the use case that one wants to make an image of 3 Gaussians and 4 Sersic profiles and 1 polynomial, i.e. different types of sources that can't be represented as one model set.
Of course one could make groups of models sets and still use the model sets feature, but then it gets complicated for the user and also implementing it would be hard ... one would need sets of extensions defining the bounding boxes where the models should be evaluated.
Even if it's possible to make this work, I think it's too complex and we shouldn't use model sets to implement this.